app/workers/investigate_botcode.py

```diff
@@ -0,0 +1,144 @@
+
+        weight_attr (str) the attribute in the edge data where the weights are.
+            in the rt graph, this represents number of times user a has retweeted user b
+    """
+
+    in_degrees = dict(rt_graph.in_degree(weight=weight_attr)) # users receiving retweets
+    out_degrees = dict(rt_graph.out_degree(weight=weight_attr)) # users doing the retweeting
+    print("IN-DEGREES:", fmt_n(len(in_degrees)))
+    print("OUT-DEGREES:", fmt_n(len(out_degrees)))
+
+    links = parse_bidirectional_links(rt_graph)
+    energies = [(link[0], link[1], compute_link_energy(link[0], link[1], link[4], in_degrees, out_degrees)) for link in links]
+    print("ENERGIES:", fmt_n(len(energies)))
+    positive_energies = [e for e in energies if sum(e[2]) > 0]
+    print("POSITIVE ENERGIES:", fmt_n(len(positive_energies)))
+
+    prior_probabilities = dict.fromkeys(list(rt_graph.nodes), 0.5)
+    energy_graph, pl, user_data = compile_energy_graph(rt_graph, prior_probabilities, positive_energies, out_degrees, in_degrees)
+    print("ENERGIES GRAPHED...") # this is the step that takes the longest
+
+    bot_probabilities = dict.fromkeys(list(user_data.keys()), 0) # start with defaults of 0 for each user
+    for user in pl:
+        user_data[user]["clustering"] = 1
+        bot_probabilities[user] = 1
+
+    return bot_probabilities, user_data
+
+if __name__ == "__main__":
+
+    graph = compile_mock_rt_graph([
+        # add some examples of users retweeting others:
+        {"user_screen_name": "user1", "retweet_user_screen_name": "leader1", "retweet_count": 40},
+        {"user_screen_name": "user2", "retweet_user_screen_name": "leader1", "retweet_count": 60},
+        {"user_screen_name": "user3", "retweet_user_screen_name": "leader2", "retweet_count": 40},
+        {"user_screen_name": "user4", "retweet_user_screen_name": "leader2", "retweet_count": 20},
+        {"user_screen_name": "user5", "retweet_user_screen_name": "leader3", "retweet_count": 40},
+        # add some examples of users retweeting eachother:
+        {"user_screen_name": "colead1", "retweet_user_screen_name": "colead2", "retweet_count": 30},
+        {"user_screen_name": "colead2", "retweet_user_screen_name": "colead1", "retweet_count": 20},
+        {"user_screen_name": "colead3", "retweet_user_screen_name": "colead4", "retweet_count": 10},
+        {"user_screen_name": "colead4", "retweet_user_screen_name": "colead3", "retweet_count": 40}
+    ])
+
+    print("----------------------")
+    in_degrees = dict(graph.in_degree(weight="rt_count")) # users receiving retweets
+    out_degrees = dict(graph.out_degree(weight="rt_count")) # users doing the retweeting
+    print("IN-DEGREES...")
+    pprint(in_degrees)
+    print("OUT-DEGREES...")
+    pprint(out_degrees)
+
+    print("----------------------")
+    print("ENSURING ALL NODES ARE REPRESENTED IN IN-DEGREE AND OUT-DEGREE VIEWS...")
+    for node in graph.nodes():
```

Comment:
todo: carry this forward as well, although an assertion might be sufficient, because all nodes seem to be represented in in and out- degrees by default.